### PR TITLE
Fix Delegator to pass keyword arguments for Ruby 3.0

### PR DESCRIPTION
**Pull request by andrewtblake**
When Ruby is upgraded to 3.0, errors result from keyword arguments not being passed correctly through the Delegator `delegate` method. This PR fixes the issue.

### CODE REVIEW

**Comment by dentarg** [comment]
Looks like this change isn't compatible with Ruby < 2.7

Would also be great to have a test for this (which fails on Ruby 3.0 until the problem has been fixed)

**Comment by andrewtblake** [comment]
Apologies for being a PR newbie.

I will work on providing a test, but maybe an informal explanation will
help in the meantime:

In a before.rb controller file, a call like

before accepted_verbs: [ 'PUT', 'POST' ] do ...

fails in Ruby 3, because the before method in sinatra base.rb L1387
receives the hash

{ accepted_verbs: ['PUT'', 'POST' ] }

in its path parameter, instead of as part of **options where it is
expecting it -- which causes the Mustermann.new method later on to complain
that it doesn't know what to do with a Hash

I believe the origin of the issue is Delegator.delegate not passing
parameters through in a way that Ruby 3 can deal with, hence the changes I
suggested in the PR

I don't really have any idea how to make my proposed changes work also in
Ruby < 2.7

**Comment by andrewtblake** [comment]
I'm not sure of the best way to provide you with my suggested test, but if
you want to try it manually it would be something like the following added
to test/delegator_test.rb starting at line 91:

====================
  it "delegates before with keyword arguments correctly" do
    delegation_app do
      set(:foo) do |something|
        something
      end
      before(foo: 'bar') do
        :nothing
      end
    end
  end

==================

With the original base.rb (before my changes) this should

   - pass on Ruby < 2.7
   - pass with a warning on Ruby >= 2.7, < 3.0
   - fail on Ruby >= 3.0, producing TypeError: Hash can't be coerced into
   Mustermann::Pattern

After my changes, the test should

   - pass on Ruby >= 2.7
   - among 72 other failures, give a weird ArgumentError on the new line 93
   of test/delegator_test.rb, when set(:foo) is called -- on Ruby < 2.7

Hope this helps

**Comment by dentarg** [comment]
@andrewtblake You can add your test from https://github.com/sinatra/sinatra/pull/1686 to this branch

From reading https://eregon.me/blog/2021/02/13/correct-delegation-in-ruby-2-27-3.html it sounds like we need to use [`ruby2_keywords`](https://github.com/ruby/ruby2_keywords). Maybe you can experiment with it?

At https://github.com/ruby/ruby2_keywords#usage there's an example that is using `define_method`

**Comment by andrewtblake** [comment]
That looks like it will do the trick, thanks. I will work on putting
together a new PR including the test and the ruby2_keywords approach.

**Comment by andrewtblake** [comment]
If you have time, I'd appreciate some guidance on where is the best place
to put the gem "ruby2_keywords", "~> 0.0.4" line in Gemfile

-- should it be inside the if RUBY_ENGINE == "ruby" block? Or above, where gem
"twitter-text", "1.14.7" is?

Also, do you prefer the require 'ruby2_keywords' line to be right above
where it is used, or alternatively at the top of the base.rb file with the
other requires?

Thanks for your help

File: lib/sinatra/base.rb
Comment (dentarg) [inline lib/sinatra/base.rb:1930]: I would put the require at the top of the file, somewhere at https://github.com/sinatra/sinatra/blob/cf404526373298f06a59e1a129b6cdf8282ae216/lib/sinatra/base.rb#L4-L10

File: Gemfile
Comment (dentarg) [inline Gemfile:69]: The Gemfile only holds our test dependencies, this will be a runtime dependency for Sinatra, so it needs to be in the gemspec: https://github.com/sinatra/sinatra/blob/cf404526373298f06a59e1a129b6cdf8282ae216/sinatra.gemspec#L46-L49

**Review by dentarg — COMMENTED** [review]
Nice work, I answered your questions about Gemfile/require

File: lib/sinatra/base.rb
Comment (dentarg) [inline lib/sinatra/base.rb:1955]: We can also avoid needing to use the `ruby2_keywords` gem if we do

```suggestion
        ruby2_keywords(method_name) if respond_to?(:ruby2_keywords, true)
```

**Comment by andrewtblake** [comment]
That seems much cleaner, although apparently the gem is already a
dependency of mustermann 1.1.1

File: lib/sinatra/base.rb
Comment (dentarg) [inline lib/sinatra/base.rb:1927]: Let's drop this newline

**Review by dentarg — APPROVED** [review]

**Comment by dentarg** [comment]
@andrewtblake if you want you can squash all commits here so it's just one so we get a nicer history

FYI I'm not a maintainer so I can't merge this

**Comment by jkowens** [comment]
Looks good to me! Thanks @andrewtblake and @dentarg 👍

**Review by epergo — APPROVED** [review]

**Review by alex-damian-negru — APPROVED** [review]

**Review by namusyaka — APPROVED** [review]

**Pull request merged**
